backends/aoti/runtime/AotiBackend.cpp

```diff
@@ -0,0 +1,532 @@
+    AOTInductorModelContainerHandle* container_handle,
+    size_t num_models,
+    const char* device_str,
+    const char* cubin_dir);
+
+  using AOTInductorModelContainerDeleteFunc = AOTIRuntimeError(*)(
+    AOTInductorModelContainerHandle container_handle);
+
+  using AOTInductorModelContainerGetNumInputsFunc = AOTIRuntimeError(*)(
+    AOTInductorModelContainerHandle container_handle,
+    size_t* num_constants);
+
+  using AOTInductorModelContainerGetNumOutputsFunc = AOTIRuntimeError(*)(
+    AOTInductorModelContainerHandle container_handle,
+    size_t* num_constants);
+
+  using AOTInductorModelContainerRunFunc = AOTIRuntimeError(*)(
+    AOTInductorModelContainerHandle container_handle,
+    AOTITensorHandle* input_handles, // array of input AOTITensorHandle; handles
+                                     // are stolen; the array itself is borrowed
+    size_t num_inputs,
+    AOTITensorHandle*
+        output_handles, // array for writing output AOTITensorHandle; handles
+                        // will be stolen by the caller; the array itself is
+                        // borrowed
+    size_t num_outputs,
+    AOTInductorStreamHandle stream_handle,
+    AOTIProxyExecutorHandle proxy_executor_handle);
+
+  AOTInductorModelContainerCreateWithDeviceFunc AOTInductorModelContainerCreateWithDevice = nullptr;
+  AOTInductorModelContainerDeleteFunc AOTInductorModelContainerDelete = nullptr;
+  AOTInductorModelContainerGetNumInputsFunc AOTInductorModelContainerGetNumInputs = nullptr;
+  AOTInductorModelContainerGetNumOutputsFunc AOTInductorModelContainerGetNumOutputs = nullptr;
+  AOTInductorModelContainerRunFunc AOTInductorModelContainerRun = nullptr;
+  std::unordered_map<Tensor*, std::vector<int64_t>> tensor_to_sizes;
+  std::unordered_map<Tensor*, std::vector<int64_t>> tensor_to_strides;
+  std::unordered_set<std::shared_ptr<Tensor>> tensors;
+
+  int32_t aoti_torch_grad_mode_is_enabled() {
+    // No autograd ever
+    return false;
+  }
+  void aoti_torch_grad_mode_set_enabled(bool enabled) {
+    if (enabled) {
+        std::runtime_error("Cannot enable autograd");
+    }
+  }
+  AOTITorchError aoti_torch_get_data_ptr(
+    AOTITensorHandle tensor,
+    void** ret_data_ptr) {
+    *ret_data_ptr = tensor->mutable_data_ptr();
+   return Error::Ok;
+  }
+  AOTITorchError aoti_torch_get_storage_offset(
+    AOTITensorHandle tensor,
+    int64_t* ret_storage_offset) {
+    // Storage offset is always 0 in ET
+    *ret_storage_offset = 0;
+    return Error::Ok;
+  }
+  AOTITorchError aoti_torch_get_strides(
+    AOTITensorHandle tensor,
+    int64_t** ret_strides) {
+    if (tensor_to_strides.find(tensor) == tensor_to_strides.end()) {
+      std::vector<int64_t> strides(tensor->dim());
+      for (int i = 0; i < tensor->dim(); i++) {
+        strides[i] = tensor->strides()[i];
+      }
+      tensor_to_strides[tensor] = strides;
+    }
+    *ret_strides = tensor_to_strides[tensor].data();
```

Comment:
you can do better with unordered_map::emplace. you should basically never use map/unordered_map's operator[];  see [Louis Brandy's excellent CppCon 2027 talk](https://www.youtube.com/watch?v=lkgszkPnV8g).

```
auto it = tensor_to_strides.find(tensor);
if (it == tensor_to_strides.end()) {
  std::vector<int64_t> strides(tensor->dim()); // should be DimVector, see comment above
  auto tensor_strides = tensor->strides();
  for (const auto i: c10::irange(tensor->dim())) {
    strides[i] = tensor_strides[i];
  }
  it = tensor_to_strides.emplace(tensor, std::move(strides)).first;
}
*ret_strides = it->data(); // here we save the additional hash table lookup compared to the original code
```